eventstore/query_json.go

```diff
@@ -0,0 +1,232 @@
+package eventstore
+
+import (
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+	"strings"
+
+	dsquery "github.com/ipfs/go-datastore/query"
+)
+
+type JSONQuery struct {
+	Ands []JSONCriterion
+	Ors  []JSONQuery
+	Sort JSONSort
+}
+type JSONCriterion struct {
+	FieldPath string
+	Operation JSONOperation
+	Value     JSONValue
+}
+
+type JSONValue struct {
+	String *string
+	Bool   *bool
+	Float  *float64
+}
+
+type JSONSort struct {
+	FieldPath string
+	Desc      bool
+}
+
+type JSONOperation int
+
+const (
+	Eq = JSONOperation(eq)
+	Ne = JSONOperation(ne)
+	Gt = JSONOperation(gt)
+	Lt = JSONOperation(lt)
+	Ge = JSONOperation(ge)
+	Le = JSONOperation(le)
+)
+
+type marshaledValue struct {
+	value   map[string]interface{}
+	rawJson []byte
+}
+
+func (t *Txn) FindJSON(q JSONQuery) ([]string, error) {
+	dsq := dsquery.Query{
+		Prefix: t.model.dsKey.String(),
+	}
+	dsr, err := t.model.store.datastore.Query(dsq)
+	if err != nil {
+		return nil, fmt.Errorf("error when internal query: %v", err)
+	}
+
+	var values []marshaledValue
+	for {
+		res, ok := dsr.NextSync()
+		if !ok {
+			break
+		}
+		val := make(map[string]interface{})
+		if err := json.Unmarshal(res.Value, &val); err != nil {
+			return nil, fmt.Errorf("error when unmarshaling query result: %v", err)
+		}
+		ok, err = q.matchJSON(val)
+		if err != nil {
+			return nil, fmt.Errorf("error when matching entry with query: %v", err)
+		}
```

Comment:
Datastore json value are unmarshaled to `map[string]interface{}`. Then the Query (`q`) is evaluated against this map.